spec/index.bs

```diff
@@ -1369,6 +1350,135 @@ interface DOMParser {
+<pre highlight="js">
+  // Assigning a string to another Realm's DOM sink uses that Realm's default policy.
+  self.top.body.innerHTML = 'Hello, World!';
+  // Using another Realm's builtin Function constructor should analogously use that
+  // Realm's default policy.
+  new self.top.Function('alert(1)')()
+</pre>
+This is subtly different from the CSP directive enforcement portion which rejects if either
+the |calleeRealm| or |callerRealm|'s Content-Security-Policy rejects string compilation.
+</div>
+
+This document modifies the grammar for [[CSP3#keyword-source]]:
+
+<pre dfn-type="grammar" link-type="grammar">
+  <dfn>keyword-source</dfn> = "'self'" / "'unsafe-inline'" / "'unsafe-eval'"
+                       / "'strict-dynamic'" / "'unsafe-hashes'" / "'report-sample'"
+                       / "'unsafe-allow-redirects'" <ins>/ "<dfn>'trusted-script'</dfn>"</ins>
+</pre>
+
+This document modifies the [[CSP3#can-compile-strings|EnsureCSPDoesNotBlockStringCompilation]]
+which is reproduced in its entirety below with additions and deletions.
+
+Given two [[ECMASCRIPT#realm|realms]] (|callerRealm| and
+|calleeRealm|), and a <del>string</del> <ins>value</ins>
+(|source|), this algorithm returns <del>normally</del>
+<ins>the source string to compile</ins> if compilation is allowed, and
+throws an "`EvalError`" if not:
+
+1.  <ins>Let |document| be |calleeRealm|’s
+    [[HTML5#environment-settings-object|environment setting object]]'s
+    [[HTML5#responsible-document|responsible document]].</ins>
+
+2.  <ins>Let |sourceString| be the result of executing the
+    [$Get Trusted Type compliant string$] algorithm, with:
+    *   |document| as |document|,
+    *   |source| as |input|,
+    *   {{TrustedScript}} as |expectedType|.</ins>
+
+3.  <ins>If the algorithm throws an error, throw an {{EvalError}}.</ins>
+
+4.  <ins>Let |isExempt| be the result of executing the [$IsSourceExempt$] algorithm with:
+    *   |document|’s {{TrustedTypeConfiguration}} as |ttConfig|.</ins>
+
+5.  Let |globals| be a list containing |calleeRealm|'s [=Realm/global object=] and |calleeRealm|'s
+    [=Realm/global object=].
+
+6.  For each |global| in |globals|:
+
+    1.  Let |result| be "`Allowed`".
+
+    2.  For each |policy| in |global|'s [[CSP#global-object-csp-list|CSP list]]:
+
+        1.  Let |source-list| be `null`.
+
+        2.  If |policy| contains a [=directive=] whose [=directive/name=] is "`script-src`", then
+            set |source-list| to that [=directive=]'s [=directive/value=].
+
+            Otherwise if |policy| contains a [=directive=] whose [=directive/name=] is
+            "`default-src`", then set |source-list| to that directive's [=directive/value=].
+
+        3.  If |source-list| is not `null`, and does not contain a [=source expression=] which is
+            an [=ASCII case-insensitive=] match for the string "<a grammar>`'unsafe-eval'`</a>", <ins>and
+            it is not the case that (|isExempt| is true and |source-list| contains an
```

Comment:
Would the language be more clear if you reverted the condition?

and (isExempt is false or source-list does not contain a match for 'trusted-script')?

There's a fair amount of double negations here :)

Reply:
So instead of

    sourceList != null && !sourceExpression.containsCaseInsensitively("'unsafe-eval'")
    && !(isExempt && sourceList.containsCaseInsensitively("'trusted-script'"))

it would be

    sourceList != null && !sourceExpression.containsCaseInsensitively("'unsafe-eval'")
    && (!isExempt || !sourceList.containsCaseInsensitively("'trusted-script'"))

which would correspond to

> If |source-list| is not `null`, and does not contain a [=source expression=] which is
> an [=ASCII case-insensitive=] match for the string "<a grammar>`'unsafe-eval'`</a>", <ins>and
> it is the case that (|isExempt| is not true or |source-list| does not contain a
> [=source expression=] which is an [=ASCII case-insensitive=] match for the
> string "<a grammar>`'trusted-script'`</a>")</ins> then

?

Reply:
Yes.

Reply:
Done.